README.md
# vformat

Please see http://hackage.haskell.org/package/vformat


ChangeLog.md
# Changelog for vformat

## V0.14.0
- Add `QuasiQuoter`s for `Format` and `Format1`

## V0.13.0
- Add `FromArgKey` typeclass
- Fix `OVERLAPPABLE` issue

## V0.12.0
- Add `defaultSpecs`
- Remove `FormatTime` instance

## V0.11.0
- Simplify `ArgKey`
- Make generic instantiate FormatArg configurable
- Expose standard formatters
- Fix some issue and doc

## V0.10.0
- Rewrite error system & fix bugs
- Fix document
- Rewrite unit tests


test/Main.hs
{-# LANGUAGE DeriveGeneric     #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Main ( main ) where

import           Data.Char
import           Data.Int
import           Data.List
import           Data.String
import           GHC.Generics
import           Numeric               hiding (showHex, showOct)
import           Test.Hspec
import           Test.Hspec.QuickCheck
import           Test.QuickCheck

import           Text.Format


main :: IO ()
main = hspec $ do
  describe "ArgKey" $ do
    it "read and show empty arg key" $
      let key = Index (-1) in show key == "" && key == read ""
    prop "read and show none empy arg key" $ \key -> checkKey key (show key)

  describe "ArgFmt" $ do
    prop "read, show and pretty" $ \fmt ->
      let raw = prettyArgFmt fmt in fmt{fmtRaw=raw} == read raw

  describe "Format" $ do
    prop "normal way" $
      uncurry (==) . formatOrdered "{0:c}{1:g}{2:g}{3:d}{4:d}{5:d}{6:s}"
    prop "all keys are omitted" $
      uncurry (==) . formatOrdered "{:c}{:g}{:g}{:d}{:d}{:d}{:s}"
    prop "some keys are omitted" $
      uncurry (==) . formatOrdered "{:c}{:g}{:g}{:d}{:d}{5:d}{6:s}"
    prop "keys are disordered" $ uncurry (==) . formatDisordered
    prop "arg format is omitted" $
      uncurry (==) . formatOrdered "{:c}{}{}{}{}{}{}"
    it "escaped brace" $
      format "{{" == ['{'] && format "}}" == ['}']
    prop "alignment and padding" $ \(align, PadChar c, n) ->
      uncurry (==) (formatAlign align c $ max 3 n)
    prop "number sign" $ \(sign, n) -> uncurry (==) (formatSign sign n)
    prop "number alternate form" $
      \(alter, n) -> uncurry (==) (formatAlter alter n)
    prop "number sign aware" $ \n -> uncurry (==) (formatSignAware n)
    prop "max width" $ \cs -> format "{:.10s}" cs == take 10 (cs :: String)
    prop "number separation" $
      \(sep, alter, n) -> uncurry (==) (formatSep sep alter n)
    prop "number precision" $
      \(NonNegative p, n) -> uncurry (==) (formatPrec p n)


instance Arbitrary ArgKey where
  arbitrary = sized $ genKey . min 10 . max 1
    where
      indexedKey = (Index . getNonNegative) <$> arbitrary
      namedKey = sized $ \n ->
        Name <$> (sequence $ replicate (max 1 n) arbitraryUnicodeChar)

      genKey 1 = oneof [indexedKey, namedKey]
      genKey n = Nest <$> (genKey 1) <*> (genKey (n - 1))


instance Arbitrary FmtAlign where
  arbitrary = elements [AlignLeft, AlignRight, AlignCenter, AlignSign]


instance Arbitrary FmtSign where
  arbitrary = elements [SignPlus, SignMinus, SignSpace]


instance Arbitrary FmtNumSep where
  arbitrary = elements [NumSepDash, NumSepComma]


instance Arbitrary ArgFmt where
  arbitrary = do
      (fmtPad, fmtAlign) <- genAlign
      fmtSign <- oneof [return SignNone, arbitrary]
      fmtAlternate <- choose (True, False)
      fmtSignAware <- choose (True, False)
      fmtWidth <- genWidth
      fmtNumSep <- oneof [return NumSepNone, arbitrary]
      fmtPrecision <- genWidth
      fmtSpecs <- getSpecs <$> arbitrary
      return $ let fmtRaw = "" in ArgFmt{..}
    where
      genAlign :: Gen (Char, FmtAlign)
      genAlign = do
        align <- oneof [return AlignNone, arbitrary]
        pad <- if align == AlignNone then return ' '
                                     else arbitraryPrintableChar
        return (pad, align)

      genWidth :: Gen (Either Int ArgKey)
      genWidth = oneof [ (Left . getNonNegative) <$> arbitrary
                       , Right <$> arbitrary
                       ]


newtype Specs = Specs { getSpecs :: String } deriving Show

instance Arbitrary Specs where
  arbitrary = do
    cs <- elements [ "", "s", "%Y-%m-%d", "b", "c", "d", "o", "x", "X"
                   , "e", "E", "f", "F", "g", "G", "%"
                   ]
    return $ Specs cs


newtype EmptyArgKey = EmptyArgKey { getEmptyArgKey :: ArgKey } deriving Show

instance Arbitrary EmptyArgKey where
  arbitrary = return $ EmptyArgKey $ Index (-1)


data BasicArgs = BasicArgs { char    :: Char
                           , double  :: Double
                           , float   :: Float
                           , integer :: Integer
                           , int     :: Int
                           , word    :: Word
                           , string  :: String
                           } deriving (Generic, Show)

instance FormatArg BasicArgs

instance Arbitrary BasicArgs where
  arbitrary = BasicArgs <$> arbitrary
                        <*> arbitrary
                        <*> arbitrary
                        <*> arbitrary
                        <*> arbitrary
                        <*> arbitrary
                        <*> arbitrary


data PadChar = PadChar { getPadChar :: Char } deriving Show

instance Arbitrary PadChar where
  arbitrary = do
    c <- suchThat arbitraryUnicodeChar $ not . (`elem` ['{', '}'])
    return $ PadChar c


data Alternate = Bin | Oct | Hex | HexU deriving (Show, Eq)

instance Arbitrary Alternate where
  arbitrary = elements [Bin, Oct, Hex, HexU]


mkSign :: (Num a, Ord a) => a -> String -> String -> String
mkSign n minus other = if n < 0 then minus else other


showBin, showOct, showHex :: Int -> String
showBin n = showIntAtBase 2 intToDigit (abs n) ""
showOct n = showIntAtBase 8 intToDigit (abs n) ""
showHex n = showIntAtBase 16 intToDigit (abs n) ""


checkKey :: ArgKey -> String -> Bool
checkKey k@(Index _) cs = show k == cs && k == read cs
checkKey k@(Name _)  cs = show k == cs && k == read cs
checkKey (Nest k1 k2) cs =
  let (cs1, cs2) = splitAt (length (show k1)) cs
  in checkKey k1 cs1 && head cs2 == '!' && checkKey k2 (tail cs2)


formatOrdered :: Format -> BasicArgs -> (String, String)
formatOrdered fs (BasicArgs{..}) =
    ( format fs char double float integer int word string
    , char : double' ++ float' ++ integer' ++ int' ++ word' ++ string
    )
  where
    double' = showGFloat (Just 6) double ""
    float' = showGFloat (Just 6) float ""
    integer' = show integer
    int' = show int
    word' = show word


formatDisordered :: BasicArgs -> (String, String)
formatDisordered (BasicArgs{..}) =
    ( format fs char double float integer int word string
    , string ++ integer' ++ float' ++ double' ++ int' ++ word' ++ [char]
    )
  where
    double' = showGFloat (Just 6) double ""
    float' = showGFloat (Just 6) float ""
    integer' = show integer
    int' = show int
    word' = show word
    fs = "{6:s}{3:d}{2:g}{1:g}{4:d}{5:d}{0:c}"


formatAlign :: FmtAlign -> Char -> Int -> (String, String)
formatAlign align c n =
    ( format (fromString $ "{:" ++ [c] ++ (mkFmt align n) ++ "}") (-42 :: Int)
    , padding align (n - 3)
    )
  where
    mkFmt :: FmtAlign -> Int -> String
    mkFmt AlignLeft   n = '<' : (show n)
    mkFmt AlignRight  n = '>' : (show n)
    mkFmt AlignCenter n = '^' : (show n)
    mkFmt AlignSign   n = '=' : (show n)

    padding AlignLeft   n = "-42" ++ replicate n c
    padding AlignRight  n = replicate n c ++ "-42"
    padding AlignCenter n =
      let ps1 = replicate (n `div` 2) c
          ps2 = replicate (n - (n `div` 2)) c
      in ps1 ++ "-42" ++ ps2
    padding AlignSign   n = "-" ++ replicate n c ++ "42"


formatSign :: FmtSign -> Int -> (String, String)
formatSign SignMinus n = (format "{:-}" n, show n)
formatSign SignPlus  n = (format "{:+}" n, (mkSign n "" "+") ++ (show n))
formatSign SignSpace n = (format "{:< d}" n, (mkSign n "" " ") ++ (show n))


formatAlter :: Alternate -> Int -> (String, String)
formatAlter Bin n = (format "{:#b}" n, (mkSign n "-" "") ++ "0b" ++ (showBin n))
formatAlter Oct n = (format "{:#o}" n, (mkSign n "-" "") ++ "0o" ++ (showOct n))
formatAlter Hex n = (format "{:#x}" n, (mkSign n "-" "") ++ "0x" ++ (showHex n))
formatAlter HexU n =
  ( format "{:#X}" n, (mkSign n "-" "") ++ "0X" ++ (map toUpper $ showHex n))


formatSignAware :: Int8 -> (String, String)
formatSignAware n =
  let cs = if n == (-128) then "128" else show $ abs n
      len = length cs + if n < 0 then 1 else 0
  in (format "{:010}" n, (mkSign n "-" "") ++ (replicate (10 - len) '0') ++ cs)


formatSep :: FmtNumSep -> Maybe Alternate -> Int -> (String, String)
formatSep sep alter n =
    ( format (fromString ("{:" ++ (mkFmt alter) ++ "}")) n
    , (mkSign n "-" "") ++ (intercalate [sepc] $ split altern $ showIt alter n)
    )
  where
    sepc = if sep == NumSepDash then '_' else ','
    altern = maybe 3 (const 4) alter

    showIt :: Maybe Alternate -> Int -> String
    showIt (Just Bin) = showBin . abs
    showIt (Just Oct) = showOct . abs
    showIt (Just _)   = showHex . abs
    showIt  _         = show . abs

    mkFmt :: Maybe Alternate -> String
    mkFmt (Just Bin) = mkFmt Nothing ++ "b"
    mkFmt (Just Oct) = mkFmt Nothing ++ "o"
    mkFmt (Just _)   = mkFmt Nothing ++ "x"
    mkFmt _          = [sepc]

    splitAtEnd :: Int -> [a] -> ([a], [a])
    splitAtEnd n cs =
      let (cs1, cs2) = splitAt n (reverse cs) in (reverse cs2, reverse cs1)

    split :: Int -> [a] -> [[a]]
    split _ [] = []
    split n cs = let (cs1, cs2) = splitAtEnd n cs in split n cs1 ++ [cs2]


formatPrec :: Int -> Double -> (String, String)
formatPrec p n =
  ( format (fromString $ "{:." ++ (if p == 0 then "" else show p) ++ "}") n
  , showGFloat (Just p) n ""
  )


test/ArgFmtSpec.hs
{-# LANGUAGE RecordWildCards #-}

module ArgFmtSpec ( spec ) where


import           Data.Maybe
import           Test.Hspec
import           Test.Hspec.QuickCheck
import           Test.QuickCheck
import           Test.QuickCheck.Monadic

import           Text.Format


spec :: Spec
spec = describe "read" $ do
  prop "pretty" $ \(TestArgFmt (cs, fmt)) -> read cs == fmt


newtype TestArgFmt = TestArgFmt (String, ArgFmt) deriving Show

instance Arbitrary TestArgFmt where
  arbitrary = do
      (fmtPad, fmtAlign) <- genAlign
      fmtSign <- genSign
      fmtAlternate <- choose (True, False)
      fmtSignAware <- choose (True, False)
      fmtWidth <- genWidth
      fmtNumSep <- elements [ NumSepNone, NumSepDash, NumSepComma ]
      fmtPrecision <- genWidth
      fmtSpecs <- genSpecs

      let fmtRaw = ""
          fmt = ArgFmt{..}
          raw = prettyArgFmt fmt

      return $ TestArgFmt (raw, fmt{fmtRaw=raw})
    where
      genAlign :: Gen (Char, FmtAlign)
      genAlign = do
        align <- elements [ AlignNone, AlignLeft, AlignRight, AlignCenter
                          , AlignSign
                          ]
        pad <- if align == AlignNone then return ' '
                                     else arbitraryPrintableChar
        return (pad, align)

      genSign :: Gen FmtSign
      genSign = elements [ SignNone, SignPlus, SignMinus, SignSpace ]

      genWidth :: Gen (Either Int ArgKey)
      genWidth = oneof [ (Left . getNonNegative) <$> arbitrary
                       , Right <$> genKey
                       ]

      genKey :: Gen ArgKey
      genKey = do
        cs <- listOf $ suchThat arbitraryPrintableChar (not . (`elem` "!{}"))
        return $ read cs

      genSpecs :: Gen String
      genSpecs = elements [ "", "s"
                          , "b", "c", "d", "o", "x", "X"
                          , "e", "E", "f", "F", "g", "G", "%"
                          , "%Y-%m-%d"
                          ]


test/ArgKeySpec.hs
module ArgKeySpec ( spec ) where


import           Test.Hspec
import           Test.Hspec.QuickCheck
import           Test.QuickCheck

import           Text.Format


spec :: Spec
spec = describe "read" $ do
  prop "empty" $ read "" == Index (-1)
  prop "index" $ \(NonNegative x) -> read (show x) == Index x
  prop "name"  $ \(NoSep cs) -> read cs == Name cs
  prop "nest"  $ \(OneSep (cs, cs1, cs2)) ->
    read cs == Nest (read cs1) (read cs2)
  prop "deep nest"  $ \(TwoSep (cs, cs1, cs2, cs3)) ->
    read cs == Nest (read cs1) (Nest (read cs2) (read cs3))
  prop "escaped" $ \(Escaped (cs, cs0)) -> read cs == Name cs0


newtype NoSep = NoSep String deriving Show

instance Arbitrary NoSep where
  arbitrary = do
    cs <- suchThat arbitrary (all (/= '!') . getNonEmpty)
    return $ NoSep $ 'x' : getNonEmpty cs


newtype OneSep = OneSep (String, String, String) deriving Show

instance Arbitrary OneSep where
  arbitrary = do
    (NoSep cs1) <- arbitrary
    (NoSep cs2) <- arbitrary
    return $ OneSep (cs1 ++ ('!' : cs2), cs1, cs2)


newtype TwoSep = TwoSep (String, String, String, String) deriving Show

instance Arbitrary TwoSep where
  arbitrary = do
    (NoSep cs1) <- arbitrary
    (NoSep cs2) <- arbitrary
    (NoSep cs3) <- arbitrary
    return $ TwoSep (cs1 ++ ('!' : cs2) ++ ('!' : cs3), cs1, cs2, cs3)


newtype Escaped = Escaped (String, String) deriving Show

instance Arbitrary Escaped where
  arbitrary = do
    (NoSep cs1) <- arbitrary
    (NoSep cs2) <- arbitrary
    return $ Escaped (cs1 ++ "!!" ++ cs2, cs1 ++ "!" ++ cs2)


src/Text/Format.hs
{-|
Module          : Text.Format
Copyright       : (c) 2019 Version Cloud
License         : BSD3
Maintainer      : Jorah Gao <jorah@version.cloud>
Stability       : experimental
Portability     : portable

This library is inspired by
Python's [str.format](https://docs.python.org/library/string.html) and
Haskell's [Text.Printf](http://hackage.haskell.org/package/base/docs/Text-Printf.html),
and most of the features are copied from these two libraries.
-}
module Text.Format
  ( -- * Format functions
    format
  , format1
    -- * Classes
  , FormatArg(formatArg)
  , FormatType(..)
  -- * Generic format arg and options
  , Options
  , defaultOptions
  , fieldLabelModifier
  , genericFormatArg
    -- * Data types
  , Formatter
  , Format
  , Format1
  , ArgKey(..)
  , topKey
  , popKey
  , ArgFmt(..)
  , prettyArgFmt
  , FmtAlign(..)
  , FmtSign(..)
  , FmtNumSep(..)
  , (:=) (..)
    -- * Standard Formatters
  , formatString
  , formatChar
  , formatInt
  , formatWord
  , formatInteger
  , formatRealFloat
  , defaultSpecs
    -- * Errors
  , ArgError(..)
  , errorArgKey
  , errorArgFmt
  , vferror
  ) where

import           Data.Map           (empty)

import           Text.Format.ArgFmt
import           Text.Format.ArgKey
import           Text.Format.Class
import           Text.Format.Error
import           Text.Format.Format


{-| Format a variable number of arguments with Python-style format string

>>> format "{:s}, {:d}, {:.4f}" "hello" 123 pi
"hello, 123, 3.1416"
>>> format "{1:s}, {0:d}, {2:.4f}" 123 "hello" pi
"hello, 123, 3.1416"
>>> format "{:s} {:d} {pi:.4f}" "hello" 123 ("pi" := pi)
"hello, 123, 3.1416"

See 'Format' to learn more about format string syntax.

See 'FormatArg' to learn how to derive FormatArg for your own data types.
-}
format :: FormatType r => Format -> r
format = flip sfmt empty

{-| A variant of 'format', it takes only one positional argument

>>> :set -XDeriveGeneric
>>> import GHC.Generics
>>> data Triple = Triple String Int Double deriving Generic
>>> instance FormatArg Triple
>>> format "{0!0:s} {0!1:d} {0!2:.4f}" $ Triple "hello" 123 pi
"hello, 123, 3.1416"
>>> format1 "{0:s} {1:d} {2:.4f}" $ Triple "hello" 123 pi
"hello, 123, 3.1416"
-}
format1 :: FormatArg a => Format1 -> a -> String
format1 = format . Format . unFormat1


src/Text/Format/Class.hs
{-# LANGUAGE DefaultSignatures    #-}
{-# LANGUAGE FlexibleContexts     #-}
{-# LANGUAGE FlexibleInstances    #-}
{-# LANGUAGE RecordWildCards      #-}
{-# LANGUAGE TypeOperators        #-}
{-# LANGUAGE UndecidableInstances #-}

module Text.Format.Class
  ( Formatter
  , FormatArg(..)
  , FromArgKey(..)
  , FormatType(..)
  , Options(..)
  , defaultOptions
  , genericFormatArg
  , (:=) (..)
  , formatString
  , formatChar
  , formatInt
  , formatWord
  , formatInteger
  , formatRealFloat
  , defaultSpecs
  ) where


import           Control.Monad.Catch
import           Data.Char
import           Data.Either
import           Data.Int
import           Data.Map            as M hiding (drop, map)
import           Data.Maybe
import           Data.Word
import           GHC.Generics
import           Numeric
import           Numeric.Natural

import           Text.Format.ArgFmt
import           Text.Format.ArgKey
import           Text.Format.Error
import           Text.Format.Format


type Formatter = ArgKey -> ArgFmt -> Either SomeException String


{-| Typeclass of formatable values.

The 'formatArg' method takes a value, a key and a field format descriptor and
either fails due to a 'ArgError' or  produce a string as the result.
There is a default 'formatArg' for 'Generic' instances, which applies
'defaultOptions' to 'genericFormatArg'.

There are two reasons may cause formatting fail

  (1) Can not find argument for the given key.

  (2) The field format descriptor does not match the argument.

==== Extending to new types

Those format functions can be extended to format types other than those
provided by default. This is done by instantiating 'FormatArg'.

Examples

@
\{\-\# LANGUAGE DeriveGeneric     \#\-\}
\{\-\# LANGUAGE OverloadedStrings \#\-\}

import           Control.Exception
import           GHC.Generics
import           Text.Format

-- Manually extend to ()
instance FormatArg () where
  formatArg x k fmt@(ArgFmt{fmtSpecs=\"U\"}) =
    let fmt' = fmt{fmtSpecs = \"\"}
    in  formatArg (show x) k fmt'
  formatArg _ _ _ = Left $ toException ArgFmtError

-- Use default generic implementation for type with nullary data constructors.
data Color = Red | Yellow | Blue deriving Generic

instance FormatArg Color

-- Use default generic implementation for type with non-nullary data constructor.
data Triple = Triple String Int Double deriving Generic

instance FormatArg Triple

-- Use default generic implementation for type using record syntax.
data Student = Student { no   :: Int
                       , name :: String
                       , age  :: Int
                       } deriving Generic

instance FormatArg Student

-- Customize field names
data Book = Book { bookName   :: String
                 , bookAuthor :: String
                 , bookPrice  :: Double
                 }

instance FormatArg Book where
  formatArg x k fmt
    | k == mempty = return $ format1 \"{name} {author} {price:.2f}\" x
    | k == Name \"name\" = formatArg (bookName x) mempty fmt
    | k == Name \"author\" = formatArg (bookAuthor x) mempty fmt
    | k == Name \"price\" = formatArg (bookPrice x) mempty fmt
    | otherwise = Left $ toException $ ArgKeyError

\-\- A better way to customize field names
\-\- instance FormatArg Book where
\-\-   formatArg = genericFormatArg $
\-\-     defaultOptions { fieldLabelModifier = drop 4 }

main :: IO ()
main = do
  putStrLn $ format \"A unit {:U}\" ()
  putStrLn $ format \"I like {}.\" Blue
  putStrLn $ format \"Triple {0!0} {0!1} {0!2}\" $ Triple \"Hello\" 123 pi
  putStrLn $ format1 \"Student: {no} {name} {age}\" $ Student 1 \"neo\" 30
  putStrLn $ format \"A book: {}\" $ Book \"Math\" \"nobody\" 99.99
  putStrLn $ format1 \"Book: {name}, Author: {author}, Price: {price:.2f}\" $
    Book \"Math\" \"nobody\" 99.99
@

Note: Since v0.12.0, 'FormatTime' instance has been remove,
use [vformat-time](http://hackage.haskell.org/package/vformat-time) instead.
-}
class FormatArg a where
  formatArg :: a -> Formatter

  default formatArg :: (Generic a, GFormatArg (Rep a)) => a -> Formatter
  formatArg = genericFormatArg defaultOptions

  formatArgList :: [a] -> Formatter
  formatArgList xs k | k == mempty = const $ throwM ArgKeyError
  formatArgList xs k | Just i <- fromArgKey (topKey k) =
    case drop (i - 1) xs of (x:_) -> formatArg x (popKey k)
  formatArgList  _ _ = const $ throwM ArgKeyError

  -- | This method is used to get the key of a top-level argument.
  -- Top-level argument means argument that directly passed to format
  -- functions ('format', 'format1').
  keyOf :: a -> ArgKey
  keyOf _ = mempty

instance FormatArg Bool

instance FormatArg Char where
  formatArg = formatChar
  formatArgList = formatString

instance FormatArg Float where
  formatArg = formatRealFloat

instance FormatArg Double where
  formatArg = formatRealFloat

instance FormatArg Int where
  formatArg = formatInt

instance FormatArg Int8 where
  formatArg = formatInt

instance FormatArg Int16 where
  formatArg = formatInt

instance FormatArg Int32 where
  formatArg = formatInt

instance FormatArg Int64 where
  formatArg = formatInt

instance FormatArg Integer where
  formatArg = formatInteger

instance FormatArg Natural where
  formatArg = formatIntegral False . toInteger

instance FormatArg Word where
  formatArg = formatWord

instance FormatArg Word8 where
  formatArg = formatWord

instance FormatArg Word16 where
  formatArg = formatWord

instance FormatArg Word32 where
  formatArg = formatWord

instance FormatArg Word64 where
  formatArg = formatWord

instance FormatArg a => FormatArg [a] where
  {-# SPECIALIZE instance FormatArg [Char] #-}
  formatArg = formatArgList

instance (Ord k, FromArgKey k, FormatArg v) => FormatArg (Map k v) where
  formatArg _ k | k == mempty = const $ throwM ArgKeyError
  formatArg x k | (Just k') <- fromArgKey (topKey k) =
    case (M.lookup k' x) of (Just x') -> formatArg x' (popKey k)
  formatArg _ _ = const $ throwM ArgKeyError


--------------------------------------------------------------------------------
{-| Typeclass for types that can be used as the key of a map-like or list-like
 container.

For example, since 'String' and 'Integral' datatypes are instance of
 'FromArgKey' and 'Double' is an instance of 'FormatArg', we can format a value
of type 'Map' 'String' 'Double'.

@since 0.13.0
-}
class FromArgKey a where
  fromArgKey :: ArgKey -> Maybe a

instance FromArgKey String where
  fromArgKey (Name k) = Just k
  fromArgKey _        = Nothing

instance Integral k => FromArgKey k where
  fromArgKey (Index i) = Just $ toEnum i
  fromArgKey _         = Nothing


--------------------------------------------------------------------------------
{-| Options that specify how to format your datatype

Options can be set using record syntax on defaultOptions with the fields below.

@since 0.11.0
-}
data Options = Options { fieldLabelModifier :: String -> String
                       }

{-| Default format options

@
'Options'
{ 'fieldLabelModifier' = id
}
@

@since 0.11.0
-}
defaultOptions = Options { fieldLabelModifier = id
                         }


{-| A configurable generic 'Formatter' creator.

@since 0.11.0
-}
genericFormatArg :: (Generic a, GFormatArg (Rep a)) => Options -> a -> Formatter
genericFormatArg opts x = gformatArg (from x) opts


--------------------------------------------------------------------------------
class GFormatArg f where
  gformatArg :: f p -> Options -> Formatter

-- Data type
instance GFormatArg f =>  GFormatArg (D1 c f) where
  gformatArg (M1 x) = gformatArg x

-- Choice between Sums
instance (GFormatArg f, GFormatArg g) => GFormatArg (f :+: g) where
  gformatArg (L1 x) = gformatArg x
  gformatArg (R1 x) = gformatArg x

-- Constructor
-- e.g. data GreetTo = Hello { name :: String } | Hi { name :: String }
--      data GreetTo = Hello String | Hi String
--      data Greet = Hello | Hi
instance (Constructor c, GFormatArg f) => GFormatArg (C1 c f) where
  gformatArg (M1 x) = gformatArg x

-- Constructor without arguments
-- e.g. data Greet = Hello | Hi
instance {-# OVERLAPPING  #-} Constructor c => GFormatArg (C1 c U1) where
  gformatArg c _ = formatArg (conName c)

-- Try Products one by one
instance (GFormatArg f, GFormatArg g) => GFormatArg (f :*: g) where
  gformatArg (x :*: y) opts k fmt =
      gformatArg x opts k fmt <|> gformatArg y opts (dec1 k) fmt
    where
      x <|> y = catchIf isArgKeyError x $ const y

      dec1 :: ArgKey -> ArgKey
      dec1 (Index i)    = Index (i - 1)
      dec1 (Nest k1 k2) = mappend (dec1 k1) k2
      dec1 k            = k

-- Selector (record and none record)
-- e.g. data GreetTo = Hello String | Hi String
--      data GreetTo = Hello { name :: String } | Hi { name :: String }
instance (Selector c, GFormatArg f) => GFormatArg (S1 c f) where
  gformatArg s@(M1 x) opts (Index 0)
    | selName s == "" = gformatArg x opts mempty
  gformatArg s@(M1 x) opts (Nest (Index 0) k)
    | selName s == "" = gformatArg x opts k
  gformatArg s@(M1 x) opts@(Options{..}) (Name record)
    | (fieldLabelModifier $ selName s) == record = gformatArg x opts mempty
  gformatArg s@(M1 x) opts@(Options{..}) (Nest (Name record) k)
    | (fieldLabelModifier $ selName s) == record = gformatArg x opts k
  gformatArg _ _ _ = const $ throwM ArgKeyError

-- FormatArg instance
instance (FormatArg c) => GFormatArg (K1 i c) where
  gformatArg (K1 c) _ = formatArg c


--------------------------------------------------------------------------------
-- | A typeclass provides the variable arguments magic for 'format'
--
class FormatType t where
  sfmt :: Format -> Map ArgKey Formatter -> t

instance (FormatArg a, FormatType r) => FormatType (a -> r) where
  sfmt fmt args = \arg -> sfmt fmt $
      insert (fixIndex $ keyOf arg) (formatArg arg) args
    where
      fixIndex k
        | k == mempty = Index $ length [n | Index n <- keys args]
        | otherwise   = k

instance FormatType String where
  sfmt fmt args = formats (unFormat fmt)
    where
      formats :: [FmtItem] -> String
      formats = concat . (map formats1)

      onError :: (ArgKey, ArgFmt) -> SomeException -> String
      onError (key, fmt) = catchArgError (errorArgKey $ show key)
                                         (errorArgFmt $ prettyArgFmt $ fmt)

      formats1 :: FmtItem -> String
      formats1 (Lit cs)       = cs
      formats1 (Arg key ifmt) = either (onError (key, ifmt)) id $
        (getFormatter key) (popKey key) (fixArgFmt ifmt)

      fixArgFmt :: ArgFmt -> ArgFmt
      fixArgFmt ifmt@(ArgFmt{fmtWidth=(Right key)}) =
        fixArgFmt $ ifmt {fmtWidth = Left $ formatWidth key}
      fixArgFmt ifmt@(ArgFmt{fmtPrecision=(Right key)}) =
        fixArgFmt $ ifmt {fmtPrecision = Left $ formatPrecision key}
      fixArgFmt ifmt = ifmt

      formatWidth, formatPrecision :: ArgKey -> Int
      formatWidth key =
        let fmt = read "0.0d"
        in read $ either (onError (key, fmt)) id $ (getFormatter key) key fmt
      formatPrecision = formatWidth

      getFormatter :: ArgKey -> Formatter
      getFormatter = maybe (\_ _ -> throwM ArgKeyError) id . (args !?) . topKey


--------------------------------------------------------------------------------
-- | A type represents the top-level named key argument.
data (:=) a = String := a
infixr 6 :=

instance FormatArg a => FormatArg ((:=) a) where
  formatArg (_ := x) k = formatArg x k
  keyOf (ks := _) = Name ks


--------------------------------------------------------------------------------
{-| Formatter for string values

@since 0.11.0
-}
formatString :: String -> Formatter
formatString _ k _ | k /= mempty = throwM ArgKeyError
formatString x _ fmt@(ArgFmt{fmtSpecs = ""})  = Right $ formatText fmt x
formatString x _ fmt@(ArgFmt{fmtSpecs = "s"}) = Right $ formatText fmt x
formatString _ _ _                            = throwM ArgFmtError


formatIntegral :: Bool -> Integer -> Formatter
formatIntegral _ _ k _ | k /= mempty = throwM ArgKeyError
formatIntegral signed x _ fmt@ArgFmt{fmtSpecs=specs} =
    formatNumber fmt signed (sepw specs) (flag specs) <$> (showx specs x)
  where
    sepw :: String -> Int
    sepw "b" = 4
    sepw "o" = 4
    sepw "x" = 4
    sepw "X" = 4
    sepw _   = 3

    flag :: String -> Maybe Char
    flag "b" = Just 'b'
    flag "o" = Just 'o'
    flag "x" = Just 'x'
    flag "X" = Just 'X'
    flag _   = Nothing

    encodeSign :: [Char] -> [Char]
    encodeSign "+" = "++"
    encodeSign "-" = "--"
    encodeSign cs  = cs

    showx :: String -> Integer -> Either SomeException String
    showx specs x
      | x < 0 = ('-' :) <$> showx specs (-x)
    showx "" x    = showx "d" x
    showx "b" x   = Right $ showIntAtBase 2 intToDigit x ""
    showx "c" x   = Right $ encodeSign $ [chr $ fromInteger x]
    showx "d" x   = Right $ show x
    showx "o" x   = Right $ showIntAtBase 8 intToDigit x ""
    showx "x" x   = Right $ showIntAtBase 16 intToDigit x ""
    showx "X" x   = map toUpper <$> showx "x" x
    showx _ _     = throwM ArgFmtError


{-| Formatter for 'Char' values

@since 0.11.0
-}
formatChar :: Char -> Formatter
formatChar = formatWord . ord

{-| Formatter for 'Int' values

@since 0.11.0
 -}
formatInt :: (Integral a, Bounded a) => a -> Formatter
formatInt = formatIntegral True . toInteger


{-| Formatter for 'Word' values

@since 0.11.0
 -}
formatWord :: (Integral a, Bounded a) => a -> Formatter
formatWord = formatIntegral False . toInteger


{-| Formatter for 'Integer' values

@since 0.11.0
 -}
formatInteger :: Integer -> Formatter
formatInteger = formatIntegral True


{-| Formatter for 'RealFloat' values

@since 0.11.0
 -}
formatRealFloat :: RealFloat a => a -> Formatter
formatRealFloat _ k _ | k /= mempty = throwM ArgKeyError
formatRealFloat x _ fmt@ArgFmt{fmtSpecs=specs, fmtPrecision=prec} =
    formatNumber fmt True 3 Nothing <$> showx specs prec1 x
  where
    prec1 = either (\i -> Just $ if i < 0 then 6 else i) (const $ Just 0) prec

    showx :: RealFloat a
          => String -> Maybe Int -> a -> Either SomeException String
    showx specs p x
      | x < 0 = ('-' :) <$> showx specs p (-x)
    showx "" p x    = showx "g" p x
    showx "e" p x   = Right $ showEFloat p x ""
    showx "E" p x   = map toUpper <$> showx "e" p x
    showx "f" p x   = Right $ showFFloat p x ""
    showx "F" p x   = map toUpper <$> showx "f" p x
    showx "g" p x   = Right $ showGFloat p x ""
    showx "G" p x   = map toUpper <$> showx "g" p x
    showx "%" p x   = (++ "%") <$> (showx "f" p (x * 100))
    showx _ _ _     = throwM ArgFmtError


{-| Use a default specs for the given formatter

@since 0.12.0
-}
defaultSpecs :: String -> (a -> Formatter) -> a -> Formatter
defaultSpecs specs f x k fmt
  | fmtSpecs fmt == "" = f x k $ fmt {fmtSpecs = specs}
  | otherwise = f x k fmt
